llvm/test/CodeGen/AMDGPU/saddo.ll

```diff
@@ -29,10 +29,11 @@ define amdgpu_kernel void @saddo_i64_zext(ptr addrspace(1) %out, i64 %a, i64 %b)
 ; SI-NEXT:    s_mov_b32 s0, s4
 ; SI-NEXT:    s_mov_b32 s1, s5
 ; SI-NEXT:    s_xor_b64 s[4:5], s[6:7], vcc
-; SI-NEXT:    v_cndmask_b32_e64 v0, 0, 1, s[4:5]
-; SI-NEXT:    v_mov_b32_e32 v1, s11
-; SI-NEXT:    v_add_i32_e32 v0, vcc, s10, v0
-; SI-NEXT:    v_addc_u32_e32 v1, vcc, 0, v1, vcc
+; SI-NEXT:    s_bfe_u64 s[4:5], s[4:5], 0x10000
+; SI-NEXT:    s_add_u32 s4, s10, s4
+; SI-NEXT:    s_addc_u32 s5, s11, s5
+; SI-NEXT:    v_mov_b32_e32 v0, s4
+; SI-NEXT:    v_mov_b32_e32 v1, s5
```

Comment:
@jayfoad on Phabricator you were wondering if this was correct (https://reviews.llvm.org/D144715)

I spent a good amount of time thinking about it and I think is correct.

`s[4:5]` is a bitmask, it contains all 1 or 0 depending on the result of the `v_cmp` earlier (both `v_cmp` have all uniform values, not sure why they're even v_cmps?). Before, we did a `v_cndmask` to put a 0 or 1 in `v0`, then did a `v_add`.

Here we just have a single bit 1/0 in the same register, and we use a `s_add`. Result is the same I believe.

Reply:
Though it's still not crystal-clear, this confuses me a bit:
```
; SI-NEXT:    s_bfe_u64 s[4:5], s[4:5], 0x10000
; SI-NEXT:    s_add_u32 s4, s10, s4
; SI-NEXT:    s_addc_u32 s5, s11, s5
```
s5 is always going to be zeroes, right? S_BFE gets one bit and will zero the other ones. That wasn't the case before I think

Reply:
> `s[4:5]` is a bitmask, it contains all 1 or 0 depending on the result of the `v_cmp` earlier

The result of v_cmp has a 0 bit for every inactive lane, so my worry is that this will not work if lane 0 happens to be inactive.

Reply:
All lanes are active here, I think it's all uniform. I think if there is a divergent block with a different exec mask, the zext wouldn't have been uniform in the first place, right?

Reply:
> All lanes are active here

You can't know that. Kernels can be launched with some lanes inactive. The set of active lanes is mostly orthogonal to which values are uniform or divergent.

Reply:
Thanks, TIL

So indeed this may be broken as-is. I guess there's one of two things we can do:

- Abandon this patch
- Make sure the `v_cmp` actually becomes a `s_cmp` so this code becomes fully scalar. I don't really know why it's using a v_cmp.

Reply:
For context the i32 cases use a v_cmp. I'm wondering if there is actually a divergence analysis bug here incorrectly marking the setcc as divergent.

Reply:
There is no divergence analysis bug, it's just that we don't have i64 scalar compares.
Would it be correct to mark a `setcc` as a source of divergence if we know it has to lower to a v_cmp?
The i1 it returns is no longer uniform, it depends on exec, right?

Reply:
No, that's not a good way to think about it. Uniformity is something the compiler can prove statically about the values in active lanes. It says nothing at all about what is happening in inactive lanes - they are outside of its domain.

Reply:
GISel still gets this one right because it marks uniforrms `G_ICMP` as using a VGPR reg bank if it knows they wont use SCC. So for the example above, GISel would use a V_CNDMASK + V_READFIRSTLANE but if the condcode is eq/ne it'll use the S_BFE form.

I'm leaning towards fixing this in SIFoldOperands instead. I'm starting to think this isn't fixable in DAGISel because we have no way to say "this is a VGPR, but it's uniform". What do you think?

Reply:
Alternative patch is #70188

Reply:
> I'm leaning towards fixing this in SIFoldOperands instead. I'm starting to think this isn't fixable in DAGISel because we have no way to say "this is a VGPR, but it's uniform". What do you think?

I really don't like adding more complexity to SIFoldOperands to fix up poor instruction selection, when it should have been possible to do better selection in the first place. But I'm not sure how to make progress on the current patch either.